pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go

```diff
@@ -670,8 +649,24 @@ func init() {
 				currentStatus(to.node, scpb.Status_PUBLIC),
 			}
 		})
+	// We need to make sure that no columns are added to the index after it
+	// receives any data due to a backfill.
+	registerDepRule("index-column added to index before index is backfilled",
+		scgraph.Precedence,
+		"index-column", "index",
+		func(from, to nodeVars) rel.Clauses {
+			return rel.Clauses{
+				from.el.Type((*scpb.IndexColumn)(nil)),
+				to.el.Type((*scpb.PrimaryIndex)(nil), (*scpb.SecondaryIndex)(nil)),
+
+				joinOnIndexID(from.el, to.el, "table-id", "index-id"),
+				targetStatusEq(from.target, to.target, scpb.ToPublic),
+				currentStatus(from.node, scpb.Status_PUBLIC),
+				currentStatus(to.node, scpb.Status_BACKFILLED),
```

Comment:
I still don't understand why this is BACKFILLED. Shouldn't this be BACKFILL_ONLY? Otherwise, won't the backfill op not see the column inside the index descriptor? Perhaps adding a test would help.

Reply:
I think this is right. We're saying we want to not allow adding any new columns to the index after the index has data. The index has data when it has become backfilled. If we were to transition the index to backfilled, by doing the backfill. `BACKFILL_ONLY` is a funny state. It effectively means, the index descriptor exists in the table descriptor, but nothing should write to it except for a backfill. However, backfill is a non-mutation operation. 

This rule doesn't matter much right now. Like I think it should exist, but it doesn't matter because we don't ever run a backfill in a transaction right now so we don't really have a risk of this not being the case. 

I suppose it's interesting to have rules that represent some property we want to be true which doesn't come up. I like having them because they express a real constraint we want the graph to have, and I hope, would prevent bad plans in a future. Maybe it's a mistake to have such rules. 

I do agree we should make sure that rules always fire. I'm working on that in a different change.